rebalance.go

```diff
@@ -0,0 +1,178 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package madmin
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type RebalStatus uint8
+
+const (
+	RebalNone RebalStatus = iota
+	RebalStarted
+	RebalCompleted
+	RebalStopped
+	RebalFailed
+)
+
+type RebalanceInfo struct {
+	StartTime time.Time   `msg:"startTs"` // Time at which rebalance-start was issued
+	EndTime   time.Time   `msg:"stopTs"`  // Time at which rebalance operation completed or rebalance-stop was called
+	Status    RebalStatus `msg:"status"`  // Current state of rebalance operation. One of Started|Stopped|Completed|Failed.
+}
+
+type RebalanceStats struct {
+	InitFreeSpace uint64 `json:"initFreeSpace"` // Pool free space at the start of rebalance
+	InitCapacity  uint64 `json:"initCapacity"`  // Pool capacity at the start of rebalance
+
+	Buckets           []string      `json:"buckets"`           // buckets being rebalanced or to be rebalanced
+	RebalancedBuckets []string      `json:"rebalancedBuckets"` // buckets rebalanced
+	Bucket            string        `json:"bucket"`            // Last rebalanced bucket
+	Object            string        `json:"object"`            // Last rebalanced object
+	NumObjects        uint64        `json:"numObjects"`        // Number of objects rebalanced
+	NumVersions       uint64        `json:"numVersions"`       // Number of versions rebalanced
+	Bytes             uint64        `json:"bytes"`             // Number of bytes rebalanced
+	Participating     bool          `json:"participating"`
+	Info              RebalanceInfo `json:"info"`
+}
+
+type RebalanceMeta struct {
+	ARN             [16]byte          `json:"arn"`             // ARN of the ongoing rebalance operation
+	PercentFreeGoal float64           `json:"percentFreeGoal"` // Computed from total free space and capacity at the start of rebalance
+	PoolStats       []*RebalanceStats `json:"poolStats"`       // Per-pool rebalance stats keyed by pool index
+	StoppedAt       time.Time         `json:"stoppedAt"`
+}
+
+func (rm *RebalanceMeta) MarshalJSON() ([]byte, error) {
+	type rebalMeta struct {
+		ARN             uuid.UUID         `json:"arn"`
+		PercentFreeGoal float64           `json:"percentFreeGoal"`
+		PoolStats       []*RebalanceStats `json:"poolStats"`
+		StoppedAt       time.Time         `json:"stoppedAt"`
+	}
+	r := rebalMeta{
+		ARN:             uuid.UUID(rm.ARN),
+		PercentFreeGoal: rm.PercentFreeGoal,
+		PoolStats:       rm.PoolStats,
+		StoppedAt:       rm.StoppedAt,
+	}
+	return json.Marshal(r)
+}
+
+func (rm *RebalanceMeta) UnmarshalJSON(b []byte) error {
+	type rebalMeta struct {
+		ARN             uuid.UUID
```

Comment:
This doesn't have json tags like MarshalJSON(), it doesn't seem this will work correctly